types/validator_set.go

```diff
@@ -545,30 +576,33 @@ func (vals *ValidatorSet) updateWithChangeSet(changes []*Validator, allowDeletes
 		return fmt.Errorf("cannot process validators with voting power 0: %v", deletes)
 	}
 
+	// Check that the resulting set will not be empty.
+	if numNewValidators(updates, vals) == 0 && len(vals.Validators) == len(deletes) {
+		return errors.New("applying the validator changes would result in empty set")
+	}
+
 	// Verify that applying the 'deletes' against 'vals' will not result in error.
-	if err := verifyRemovals(deletes, vals); err != nil {
+	// Get the voting power that is going to be removed.
+	removedVotingPower, err := verifyRemovals(deletes, vals)
+	if err != nil {
 		return err
 	}
 
 	// Verify that applying the 'updates' against 'vals' will not result in error.
-	updatedTotalVotingPower, numNewValidators, err := verifyUpdates(updates, vals)
+	// Get the updated total voting power before removal. Note that this is < 2 * MaxTotalVotingPower
+	tvpAfterUpdatesBeforeRemovals, err := verifyUpdates(updates, vals, removedVotingPower)
 	if err != nil {
 		return err
 	}
 
-	// Check that the resulting set will not be empty.
-	if numNewValidators == 0 && len(vals.Validators) == len(deletes) {
-		return errors.New("applying the validator changes would result in empty set")
-	}
-
 	// Compute the priorities for updates.
-	computeNewPriorities(updates, vals, updatedTotalVotingPower)
+	computeNewPriorities(updates, vals, tvpAfterUpdatesBeforeRemovals)
```

Comment:
I'm confused about why we're passing in the tvp before the removals now, especially since it looks like the implementation of computeNewPriorities hasn't changed?

Reply:
I think the code there will benefit from refactoring (example: https://github.com/tendermint/tendermint/pull/4183/commits/02a89fca9457738550173efe0517c9a139a41f69)

Reply:
Can you elaborate more on that example? Are you saying we should use a switch here too? Or just giving that as an example of some effective refactoring in the same part of the codebase?

Reply:
The latter 😄

Reply:
> I'm confused about why we're passing in the tvp before the removals now, especially since it looks like the implementation of computeNewPriorities hasn't changed?

It was the same before, just changed the variable name.

Reply:
Ahh, I see. I couldn't tell at first from reading the guts of `verifyUpdates`, but this makes sense. Thanks!